File: android/app/src/main/java/com/dailymemory/domain/usecase/memory/DeleteMemoryUseCase.kt
package com.dailymemory.domain.usecase.memory

import com.dailymemory.domain.repository.MemoryRepository
import com.dailymemory.domain.repository.ReminderRepository
import kotlinx.coroutines.flow.first
import javax.inject.Inject

/**
 * Use case for deleting a memory.
 * Also handles cleanup of related reminders.
 */
class DeleteMemoryUseCase @Inject constructor(
    private val memoryRepository: MemoryRepository,
    private val reminderRepository: ReminderRepository
) {
    suspend operator fun invoke(memoryId: String): Result<Unit> {
        return try {
            // Delete related reminders first
            val relatedReminders = reminderRepository.observeByMemoryId(memoryId).first()
            relatedReminders.forEach { reminder ->
                reminderRepository.deleteById(reminder.id)
            }

            // Delete the memory
            memoryRepository.deleteById(memoryId)

            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

File: ios/DailyMemory/Domain/UseCase/Memory/DeleteMemoryUseCase.swift
import Foundation

/// Use case for deleting a memory.
/// Also handles cleanup of related reminders.
final class DeleteMemoryUseCase {
    private let memoryRepository: MemoryRepository
    private let reminderRepository: ReminderRepository

    init(
        memoryRepository: MemoryRepository = MemoryRepositoryImpl(),
        reminderRepository: ReminderRepository = ReminderRepositoryImpl()
    ) {
        self.memoryRepository = memoryRepository
        self.reminderRepository = reminderRepository
    }

    func execute(memoryId: String) async throws {
        // Delete related reminders first
        let relatedReminders = try await reminderRepository.getByMemoryId(memoryId)
        for reminder in relatedReminders {
            try await reminderRepository.delete(id: reminder.id)
        }

        // Delete the memory
        try await memoryRepository.delete(id: memoryId)
    }
}
